source/simulator/helper_functions.cc

```diff
@@ -2403,6 +2404,235 @@ namespace aspect
+          MaterialModel::MaterialModelInputs<dim> in(fe_values, cell, introspection, solution);
+          MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, introspection.n_compositional_fields);
+          material_model->evaluate(in, out);
+
+          MaterialModel::MaterialModelInputs<dim> in_adjoint(fe_values, cell, introspection, current_adjoint_solution);
+
+          local_rhs = 0.;
+          local_mass_matrix = 0.;
+
+
+          // VOLUME TERM
+
+          for (unsigned int q=0; q<n_q_points; ++q)
+            {
+              // Get parameters for the density kernel
+              const Tensor<1,dim> velocity_adjoint = in_adjoint.velocity[q];
+              const Tensor<1,dim> gravity = gravity_model->gravity_vector(in.position[q]);
+
+              // Get parameters for the viscosity kernel
+              const double eta = out.viscosities[q];
+              const SymmetricTensor<2,dim> strain_rate_forward = in.strain_rate[q] - 1./3 * trace(in.strain_rate[q]) * unit_symmetric_tensor<dim>();
+              const SymmetricTensor<2,dim> strain_rate_adjoint = in_adjoint.strain_rate[q] - 1./3 * trace(in_adjoint.strain_rate[q]) * unit_symmetric_tensor<dim>();
+
+              for (unsigned int k=0; k<dofs_per_cell; ++k)
+                {
+                  phi_rho[k] = fe_values[introspection.extractors.compositional_fields[0]].value(k,q);
+                  phi_eta[k] = fe_values[introspection.extractors.compositional_fields[1]].value(k,q);
+                }
+
+              for (unsigned int i = 0; i<dofs_per_cell; ++i)
+                {
+                  // Calculate the density kernel
+                  local_rhs(i) += -(gravity * velocity_adjoint) * phi_rho[i] * fe_values.JxW(q);
+
+                  // Calculate the viscosity kernel
+                  local_rhs(i) += (2 * eta * strain_rate_forward * strain_rate_adjoint) * phi_eta[i] * fe_values.JxW(q);
+                  // it's fine to add both of them because only one of them at a time is nonzero
+
+                  for (unsigned int j = 0; j<dofs_per_cell; j++)
+                    {
+                      // Assemble Mass Matrix for the first compositional field. the mass
+                      // matrix for the second compositional field would look exactly the
+                      // same, so we can avoid building it for efficiency. (in fact we
+                      // also would have to deal with the fact that no sparsity pattern
+                      // and memory is allocated for the second compositional field; but
+                      // because we don't build the matrix, we don't care here.)
+                      local_mass_matrix(i,j) += fe_values[introspection.extractors.compositional_fields[0]].value(i,q) *
+                                                fe_values[introspection.extractors.compositional_fields[0]].value(j,q) *
+                                                fe_values.JxW(q);
```

Comment:
Isnt this the same as:

```suggestion
                      local_mass_matrix(i,j) += phi_rho[i] *
                                                phi_rho[j] *
                                                JxW;
```

This would be much faster, because the three quantities would not need to be computed n_dof * n_dof times.